### PR TITLE
feat(multiselect): Add the ability to select multiple dates

### PR DESCRIPTION
**Pull request by isaacplmann**
For issue #11.

`dp-calendar` expects the `selected` input to always be a `Moment[]`.
`dp-day-picker` handles `ngModel` input as `string | Moment | Moment[]`.
 
The value that is returned by `ngModelChange` corresponds to the `userValue` that was passed in.

Once I finish my check list, I don't think this will cause any breaking changes.

I still need to:

- [x] disable multiselect by default
- [x] write better tests to ensure that the output is always in the format you pass in
- [x] squash commits into one commit

There are 2 ways I could control whether multiselect is enabled:
1. Check to see if the initial `userValue` is an array (to get multiselect with no initial value users will need to pass `[]` instead of `undefined`)
2. Add an `allowMultiSelect` boolean value on the `IDayPickerConfig`.

Also, should I work on `minSelections` and `maxSelections` in this PR?  Those would take the place of `allowMultiSelect`.

### CODE REVIEW

**Comment by isaacplmann** [comment]
Also, the `_value` in `dp-day-picker` is always `Moment[]`, the conversion back to the user format is done in the getter and setter.  I'm referencing `_value` in the template of `dp-day-picker` to ensure that the value passed to `dp-calendar` is always `Moment[]`.  Not sure if that's the best way to do it, but I couldn't think of a cleaner way.  Suggestions welcome.

**Comment by isaacplmann** [comment]
And another thorny issue:

It's hard to differentiate between `string[]` and `Moment[]` user input.  Right now I'm handling comma separated string values as multiple dates and returning the value back as csv.  If we want the user to be able to pass in a `string[]` value, I think we'll need to add an explicit `userValueType` config option.

File: src/app/dp-day-picker/dp-day-picker.component.ts [hunk 1]
```diff
@@ -1,12 +1,5 @@
-import {
-  Component,
-  Input,
-  HostListener,
-  forwardRef,
-  SimpleChanges,
-  OnChanges,
-  OnInit
-} from '@angular/core';
+import { ICalendarDay } from '../dp-calendar/config/day.model';
```
Comment (vlio20) [inline src/app/dp-day-picker/dp-day-picker.component.ts]: please remove spaces in import { ... } -> {...}

File: src/app/dp-day-picker/dp-day-picker.component.ts [hunk 2]
```diff
@@ -1,12 +1,5 @@
...
+import { Component, forwardRef, HostListener, Input, OnChanges, OnInit, SimpleChanges } from '@angular/core';
```
Comment (vlio20) [inline src/app/dp-day-picker/dp-day-picker.component.ts]: same

**Comment by vlio20** [comment]
Please add the following:  
1. Add the `allowMultiSelect` option to the configuration - users should explicitly ask for this functionality.
2. Add this to documentation.
3. Please publish an example with your changes (plunker or publish a demo).

thanks

**Comment by isaacplmann** [comment]
do you want `allowMultiSelect` or `minSelections` and `maxSelections`

**Comment by isaacplmann** [comment]
I'm a lot more confident now.

1. `_value` is not referenced anywhere outside the getter and setter.
2. `dp-day-picker`'s `value` is always of type `Moment[]`, but the `processOnChangeCallback` function returns the value in the format the user sets (implicitly or explicitly)
3. There is an `allowMultiSelect` option that defaults to false (but will become true if `allowMultiSelect` is undefined and the user passes in an array)
4. There is a `userValueType` option where a user can explicitly say what format they want the value in.  (This is specifically needed when a user passes in a `string[]`, but we can't differentiate that from a `Moment[]`.)
5. The demo page now allows you to set `allowMultiSelect` and `userValueType`
6. The demo page also has a link at the bottom to a more extensive Multiselect Test page.

Lots of code.  I tried to keep me coding style consistent with yours.  Let me know what you think.

File: src/app/demo/demo/demo.component.html [hunk 3]
```diff
@@ -7,7 +7,7 @@
       Installation guide can be found in the repo README.md file.
     </p>
     <p>
-      Fill free to open issues, suggest improvements and submit PRs.
+      Feel free to open issues, suggest improvements and submit PRs.
```
Comment (vlio20) [inline src/app/demo/demo/demo.component.html]: :)

**Comment by vlio20** [comment]
I will play with it and approve if all seems ok.
Will update soon

File: src/app/demo/demo.module.ts [hunk 4]
```diff
@@ -1,19 +1,38 @@
+import { DpDayPickerComponent } from './../dp-day-picker/dp-day-picker.component';
```
Comment (vlio20) [inline src/app/demo/demo.module.ts]: remove spaces please

File: src/app/demo/demo/demo.component.ts [hunk 5]
```diff
@@ -28,8 +28,10 @@ export class DemoComponent {
     calendarsAmount: 1,
     format: 'DD-MM-YYYY',
     monthFormat: 'MMM, YYYY',
+    // allowMultiSelect: true,
```
Comment (vlio20) [inline src/app/demo/demo/demo.component.ts]: can this be removed?

File: src/app/dp-day-picker/dp-day-picker.component.ts [hunk 6]
```diff
@@ -108,12 +99,22 @@ export class DpDayPickerComponent implements OnChanges, OnInit, ControlValueAcce
 
   writeValue(value: Moment): void {
     if (value) {
-      this.userValueType = typeof value;
+      this.pickerConfig.userValueType = this.pickerConfig.userValueType || <any>typeof value;
```
Comment (vlio20) [inline src/app/dp-day-picker/dp-day-picker.component.ts]: is there a way to give this any casting a type?

File: src/app/dp-day-picker/dp-day-picker.component.ts [hunk 7]
```diff
@@ -108,12 +99,22 @@ export class DpDayPickerComponent implements OnChanges, OnInit, ControlValueAcce
...
       this.userValue = value;
       this.init();
     }
   }
 
+  processOnChangeCallback(value: Moment[]): string | string[] | Moment | Moment[] {
```
Comment (vlio20) [inline src/app/dp-day-picker/dp-day-picker.component.ts]: maybe we need to define a type to all the available value type options:
```
CalendarValue: string | string[] | Moment | Moment[]
```

File: src/app/dp-day-picker/dp-day-picker.component.ts (hunk 6 above)
Comment (isaacplmann) [inline src/app/dp-day-picker/dp-day-picker.component.ts]: I'll explicitly check if it's equal to 'string'

Comment (vlio20) [inline src/app/dp-day-picker/dp-day-picker.component.ts]: cool

**Comment by isaacplmann** [comment]
Added documentation in the Readme.md.
Added a `DateLike` type.

**Comment by isaacplmann** [comment]
Just saw you wanted it to be named `CalendarValue` instead of `DateLike`

**Comment by vlio20** [comment]
It doesn't really matter. 
Can you please address the other comments?

**Comment by isaacplmann** [comment]
I think I did.  Can you refresh and check?

**Comment by vlio20** [comment]
@isaacplmann, thanks a lot!!!

**Pull request merged**